src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaConsumerConsumeIntegration.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Kafka
+{
+    /// <summary>
+    /// Confluent.Kafka Consumer.Consume calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Confluent.Kafka",
+        TypeName = "Confluent.Kafka.Consumer`2",
+        MethodName = "Consume",
+        ReturnTypeName = KafkaConstants.ConsumeResultTypeName,
+        ParameterTypeNames = new[] { ClrNames.Int32 },
+        MinimumVersion = "1.4.0",
+        MaximumVersion = "1.*.*",
+        IntegrationName = KafkaConstants.IntegrationName)]
+    public class KafkaConsumerConsumeIntegration
+    {
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="millisecondsTimeout">The maximum period of time the call may block.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget>(TTarget instance, int millisecondsTimeout)
+        {
+            // If we are already in a consumer scope, close it, and start a new one on method exit.
+            KafkaHelper.CloseConsumerScope(Tracer.Instance);
+            return CallTargetState.GetDefault();
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TResponse">Type of the response</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="response">Response instance</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        public static CallTargetReturn<TResponse> OnMethodEnd<TTarget, TResponse>(TTarget instance, TResponse response, Exception exception, CallTargetState state)
+        {
+            IConsumeResult consumeResult = null;
+            if (response is not null)
+            {
+                // Have to do the cast here, not in the method signature, as otherwise response is never null
```

Comment:
If you add the generic constraint `TResponse : IConsumeResult, IDucktype`, then you could do your "null check" by checking if `response.Instance is not null`. I'm not sure if one way or the other is faster, but at least that way you wouldn't have to go out of your way to do the cast

Reply:
Ahhh, I was trying this before without the `IDucktype` interface, and was running into issues because `response` is never null! Hadn't noticed `IDucktype` before, I'll update that, thanks 👍

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/1444/commits/277945283c59bd5569a1d00d9904bd322675bc8c